File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/views/CurrencyViews.java
package com.habitrpg.android.habitica.ui.views;

import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.widget.LinearLayout;

import com.habitrpg.android.habitica.R;

/**
 * Created by phillip on 08.09.17.
 */

public class CurrencyViews extends LinearLayout {
    private CurrencyView hourglassTextView;
    private CurrencyView goldTextView;
    private CurrencyView gemTextView;
    public boolean lightbackground;

    public CurrencyViews(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        TypedArray attributes = context.getTheme().obtainStyledAttributes(
                attrs,
                R.styleable.CurrencyViews,
                0, 0);

        lightbackground = attributes.getBoolean(R.styleable.CurrencyViews_hasLightBackground, true);

        setupViews();
    }

    public CurrencyViews(Context context) {
        super(context);
        setupViews();
    }

    private void setupViews() {
        Resources r = getContext().getResources();
        int margin = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 12, r.getDisplayMetrics());

        hourglassTextView = new CurrencyView(getContext(), "hourglasses", lightbackground);
        this.addView(hourglassTextView);
        LinearLayout.LayoutParams hourglassParams = (LayoutParams) hourglassTextView.getLayoutParams();
        hourglassParams.setMargins(margin, 0, 0, 0);
        hourglassTextView.setLayoutParams(hourglassParams);
        goldTextView = new CurrencyView(getContext(), "gold", lightbackground);
        this.addView(goldTextView);
        LinearLayout.LayoutParams goldParams = (LayoutParams) goldTextView.getLayoutParams();
        goldParams.setMargins(margin, 0, 0, 0);
        goldTextView.setLayoutParams(goldParams);
        gemTextView = new CurrencyView(getContext(), "gems", lightbackground);
        this.addView(gemTextView);
        LinearLayout.LayoutParams gemParams = (LayoutParams) gemTextView.getLayoutParams();
        gemParams.setMargins(margin, 0, 0, 0);
        gemTextView.setLayoutParams(gemParams);
    }

    public void setGold(Double gold) {
        goldTextView.setValue(gold);
    }

    public void setGems(Integer gemCount) {
        gemTextView.setValue(Double.valueOf(gemCount));
    }

    public void setHourglasses(Integer hourglassCount) {
        hourglassTextView.setText(String.valueOf(hourglassCount));
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/views/CurrencyViews.kt
package com.habitrpg.android.habitica.ui.views

import android.content.Context
import android.util.AttributeSet
import android.util.TypedValue
import android.widget.LinearLayout
import com.habitrpg.android.habitica.R
import com.habitrpg.common.habitica.extensions.isUsingNightModeResources

class CurrencyViews : LinearLayout {
    var lightBackground: Boolean = false
        set(value) {
            field = value
            hourglassTextView.lightBackground = value
            gemTextView.lightBackground = value
            goldTextView.lightBackground = value
        }
    private val hourglassTextView: CurrencyView = CurrencyView(context, "hourglasses", lightBackground)
    private val goldTextView: CurrencyView = CurrencyView(context, "gold", lightBackground)
    private val gemTextView: CurrencyView = CurrencyView(context, "gems", lightBackground)

    var gold: Double
        get() = goldTextView.value
        set(value) { goldTextView.value = value }
    var gems: Double
        get() = gemTextView.value
        set(value) { gemTextView.value = value }
    var hourglasses: Double
        get() = hourglassTextView.value
        set(value) { hourglassTextView.value = value }

    var hourglassVisibility
        get() = hourglassTextView.visibility
        set(value) {
            hourglassTextView.visibility = value
            hourglassTextView.hideWhenEmpty = false
        }
    var goldVisibility: Int
        get() = goldTextView.visibility
        set(value) { goldTextView.visibility = value }
    var gemVisibility
        get() = gemTextView.visibility
        set(value) { gemTextView.visibility = value }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        val attributes = context.theme?.obtainStyledAttributes(
            attrs,
            R.styleable.CurrencyViews,
            0, 0
        )
        setupViews()
        val fallBackLight = !context.isUsingNightModeResources()
        lightBackground = attributes?.getBoolean(R.styleable.CurrencyViews_hasLightBackground, fallBackLight) ?: fallBackLight
    }

    constructor(context: Context?) : super(context) {
        setupViews()
    }

    private fun setupViews() {
        val margin = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 12f, context.resources.displayMetrics).toInt()
        setupView(hourglassTextView, margin)
        setupView(goldTextView, margin)
        setupView(gemTextView, margin)
    }

    private fun setupView(view: CurrencyView, margin: Int) {
        this.addView(view)
        view.textSize = 12f
        val params = view.layoutParams as? LayoutParams
        params?.marginStart = margin
        view.layoutParams = params
    }
}
